### PR TITLE
NEW Refactor CLI interaction with Silverstripe app

### PR DESCRIPTION
**Pull request by GuySartorelli**
- Turn sake into a symfony/console app
- Avoid using HTTPRequest for CLI interaction
- Implement abstracted (named "poly" in these PRs) execution path for CLI and HTTP code paths to land in the same place

The idea is that you can define any combination of these in a module or project:
- pure symfony/console commands for CLI-only interaction
- pure controllers for HTTP-only interaction (but still in the `dev/*` route) e.g. [`/dev/graphql/ide`](https://github.com/silverstripe/silverstripe-graphql-devtools/blob/1/src/Controller.php) which simply doesn't work in a CLI context
- "poly" functionality that can be executed from either a CLI context or a HTTP context - but which doesn't need to duplicate all it's output logic based on `Director::is_cli()`
    - Most `/dev/*` especially `BuildTask` subclasses will use this
    - Can also give explicit HTTP routes via director routing rules and add like normal commands for CLI

New sake uses `tasks:mytask` format in CLI (this is standard for symfony console apps and means we can take advantage of grouping functionality), but the old `dev/tasks/mytask` syntax will still work for at least the lifetime of CMS 6 so people don't have to immediately rewrite all their scripts. The `dev/` aliases are deprecated to be removed in the future.

Flags can also still be passed as arguments (e.g. `--flush` can still be passed as `flush=1`) but this is deprecated.

The HTTP `/dev/*` routes (e.g. `/dev/build`) will still work exactly the same way they used to, from a user perspective. The only difference is where the code lives, how it's registered, and how the `DevelopmentAdmin` class calls it - and the fact that additional help info including available parameters are shown in the list views.
I'm pretty sure this is a low-traffic area in terms of customisation but the upgrade path is very straight forward in any case.

Existing `BuildTask` implementations which `echo` out their output directly should still work effectively as-is (i.e. just update the method and property signatures but don't update the method body) if people don't want to update to use the new `PolyOutput`.
The main thing that will change is that you won't have an `HTTPRequest` anymore - instead get vars and CLI flags will be given via symfony's `InputInterface`. Should be a very straight forward upgrade.

I've implemented a `navigate` command (e.g. `sake navigate about-us`) in case people have weird edge cases where they do explicitly want to use the old HTTPRequest-over-CLI style. But its use will be discouraged in favour of implementing a `PolyCommand`.

## Issue
- https://github.com/silverstripe/silverstripe-framework/issues/11341

### CODE REVIEW

File: src/Dev/HybridExecution/HybridOutput.php
Comment (GuySartorelli) [inline src/Dev/HybridExecution/HybridOutput.php:1]: I need to consider whether this should be given its context on instantiate by passing it into the constructor (which is the current way I'm doing it), or whether it should just check `Director::is_cli()` and base its context on that.

The latter would be simpler... I just need to consider if there's a use case for explicitly saying "I want CLI output" even when you're in a web request context.

One scenario might be if you're queuing a task, and then running the queue via the CMS. I'll have to double check but I'm pretty sure `Director::is_cli()` would return true in that scenario, but we don't want HTML in the output 'cause it doesn't get rendering in the queued job message log.

Comment (GuySartorelli) [reply src/Dev/HybridExecution/HybridOutput.php:1]: I've decided to stick with passing in the context as a constructor argument, because:
1. It's a little easier to test
2. It allows use cases like queued jobs running tasks to get the CLI-style output instead of HTML even when run from an actual HTTP request

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Change `Hybrid*` to `Poly*`

Reason being that `Hybird` means AND whereas `Poly` is OR - e.g. polymorphic

File: _config/confirmation-middleware.yml [hunk 1]
```diff
@@ -22,14 +22,10 @@ SilverStripe\Core\Injector\Injector:
     class: SilverStripe\Control\Middleware\ConfirmationMiddleware\EnvironmentBypass
     type: prototype
 
-  SilverStripe\Control\Middleware\ConfirmationMiddleware\CliBypass:
-    class: SilverStripe\Control\Middleware\ConfirmationMiddleware\CliBypass
-    type: prototype
```
Comment (GuySartorelli) [inline _config/confirmation-middleware.yml:27]: Deleted since CLI doesn't use the HTTP routing (and therefore doesn't use the middleware) anymore

File: _config/logging.yml [hunk 2]
```diff
@@ -52,7 +52,7 @@ Only:
 # Dev handler outputs detailed information including notices
 SilverStripe\Core\Injector\Injector:
   Monolog\Handler\HandlerInterface:
-    class: SilverStripe\Logging\HTTPOutputHandler
+    class: SilverStripe\Logging\ErrorOutputHandler
```
Comment (GuySartorelli) [inline _config/logging.yml:55]: That got renamed because it's not _only_ used for `HTTP`.

File: client/styles/debug.css [hunk 3]
```diff
@@ -128,22 +127,28 @@ a:active {
     line-height: 1.3;
 }
 
-.build .success {
+.options .success {
```
Comment (GuySartorelli) [inline client/styles/debug.css:130]: `build` was used for dev/build but also for one or two other random things. Now these styles are available for all dev/* HTTP requests.

File: src/Cli/LegacyParamArgvInput.php [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+ *
+ * @deprecated 6.0.0 Use Symfony\Component\Console\Input\ArgvInput instead.
+ */
+class LegacyParamArgvInput extends ArgvInput
+{
+    /**
+     * Input from the command line.
+     *
+     * We need a separate copy of this because the one held by the parent class is private
+     * and not exposed until symfony/console 7.1
```
Comment (GuySartorelli) [inline src/Cli/LegacyParamArgvInput.php:25]: Didn't realise 7.1 had it exposed until I'd already implemented all this - this all works fine, and is explicitly deprecated code, and allows us to keep all of the symfony deps at `7.0` instead of having one outlier.
Using the new API in 7.1 would change _very_ little so probably not worth upgrading to 7.1 just for that.

File: src/Cli/Sake.php [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+            }
+        }
+    }
+
+    public function all(?string $namespace = null): array
+    {
+        $commands = parent::all($namespace);
+        // If number of tasks is greater than the limit, hide them from the main comands list.
+        $maxTasks = Sake::config()->get('max_tasks_to_display');
+        if (!$this->ignoreTaskLimit && $maxTasks > 0 && $namespace === null) {
```
Comment (GuySartorelli) [inline src/Cli/Sake.php:130]: Not using `shouldHideTasks` here because that calls `all()`.
That method is for use _outside_ `all()`.

File: src/Cli/Sake.php [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+            $toHide[] = DumpCompletionCommand::class;
+        }
+        foreach ($commands as $command) {
+            if (in_array(get_class($command), $toHide)) {
+                $command->setHidden(true);
+            }
+        }
+
+        $commands[] = $this->createFlushCommand();
+        $commands[] = new TasksCommand();
```
Comment (GuySartorelli) [inline src/Cli/Sake.php:249]: Added here instead of `Sake.Commands` config because:

- Flush is really just a shell, it's not doing anything. Allowing dependency injection for that suggests more is going on in that command than really is.
- The tasks command does one specific thing - tell the list command to list tasks. There's no utility in overriding it, and we do not want people to remove it.

File: src/Control/CLIRequestBuilder.php [hunk 7]
```diff
@@ -48,12 +49,12 @@ public static function cleanEnvironment(array $variables)
         if (isset($variables['_SERVER']['argv'][2])) {
             $args = array_slice($variables['_SERVER']['argv'] ?? [], 2);
             foreach ($args as $arg) {
-                if (strpos($arg ?? '', '=') == false) {
+                if (strpos($arg ?? '', '=') === false) {
```
Comment (GuySartorelli) [inline src/Control/CLIRequestBuilder.php:52]: `0` used to be caught here, but it shouldn't.

File: src/Control/CLIRequestBuilder.php [hunk 8]
```diff
@@ -48,12 +49,12 @@ public static function cleanEnvironment(array $variables)
...
                     $variables['_GET']['args'][] = $arg;
                 } else {
                     $newItems = [];
                     parse_str((substr($arg ?? '', 0, 2) == '--') ? substr($arg, 2) : $arg, $newItems);
-                    $variables['_GET'] = array_merge($variables['_GET'], $newItems);
+                    $variables['_GET'] = array_merge_recursive($variables['_GET'], $newItems);
```
Comment (GuySartorelli) [inline src/Control/CLIRequestBuilder.php:57]: This allows `val[]=1 val[]=2` syntax for array getvars.

File: src/Control/CliController.php
Comment (GuySartorelli) [inline src/Control/CliController.php:1]: This wasn't used, and we don't want it now that Sake isn't using the director routes as its primary way of executing code.

File: src/Core/Kernel.php [hunk 9]
```diff
@@ -139,4 +139,9 @@ public function setEnvironment($environment);
      * @return bool|null null if the kernel hasn't been booted yet
      */
     public function isFlushed(): ?bool;
+
+    /**
+     * Returns whether the kernel has been booted
+     */
+    public function getBooted(): bool;
```
Comment (GuySartorelli) [inline src/Core/Kernel.php:146]: This was added to `CoreKernel` - makes sense to add to the interface.

File: src/Dev/Command/ConfigAudit.php
Comment (GuySartorelli) [inline src/Dev/Command/ConfigAudit.php:1]: This was a "hidden" command available on the controller that provided `dev/config`.
You'll notice a pattern that anything that was available on `dev/*` but was hidden is now being explicitly exposed. There was no sense in hiding things that devs may find useful.

I won't comment on them all - but just anything that's in the `/dev` list now that wasn't in that list in CMS 5 - you could already run it, it just wasn't in the list before.

File: src/Dev/Command/GenerateSecureToken.php
Comment (GuySartorelli) [inline src/Dev/Command/GenerateSecureToken.php:1]: This is a bit of a weird one - the old `/dev/generatesecuretoken` seemed like it was for some specific now-removed functionality, given the YAML snippet it would output.

I've removed the YAML snippet and made the token algorithm configurable, but decided to otherwise keep this as it may be useful.

File: src/Dev/MigrationTask.php [hunk 10]
```diff
@@ -2,77 +2,49 @@
...
-    }
+    /**
+     * Migrate from old to new
+     */
+    abstract public function up();
+
+    /**
+     * Revert the migration (new to old)
+     */
+    abstract public function down();
```
Comment (GuySartorelli) [inline src/Dev/MigrationTask.php:36]: These didn't used to be abstract - but the `execute()` method expects them to do something.
This class has no utility if these two methods are not implemented.

I considered just removing the class but if we want to do that it should be a separate card where people can comment if they actually find it useful and want it to stick around.

File: src/Dev/Tasks/i18nTextCollectorTask.php [hunk 11]
```diff
@@ -2,83 +2,71 @@
...
 use SilverStripe\i18n\TextCollection\i18nTextCollector;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
 
 /**
  * Collects i18n strings
+ *
+ * It will search for existent modules that use the i18n feature, parse the _t() calls
+ * and write the resultant files in the lang folder of each module.
```
Comment (GuySartorelli) [inline src/Dev/Tasks/i18nTextCollectorTask.php:18]: Taken from the phpdoc of the old `run()` method.

File: src/Logging/ErrorOutputHandler.php
Comment (GuySartorelli) [inline src/Logging/ErrorOutputHandler.php:1]: More may need to be done to improve the error handling for CLI - I'll create a separate follow-up card for that.

File: src/PolyExecution/AnsiToHtmlConverter.php [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        }
+
+        // These lines commented out from the parent class implementation.
+        // We don't want this opinionated default colouring - it doesn't appear in the ANSI format so it doesn't belong in the output.
+        // if ($this->inlineStyles) {
+        //     $html = sprintf('<span style="background-color: %s; color: %s">%s</span>', $this->inlineColors['black'], $this->inlineColors['white'], $html);
+        // } else {
+        //     $html = sprintf('<span class="ansi_color_bg_black ansi_color_fg_white">%s</span>', $html);
+        // }
+        // We do need an opening and closing span though, or the HTML markup is broken
```
Comment (GuySartorelli) [inline src/PolyExecution/AnsiToHtmlConverter.php:69]: I've commented that out so it's clear what has changed from the superclass - but happy to remove if you want it removed.

File: src/PolyExecution/AnsiToHtmlTheme.php [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+ */
+class AnsiToHtmlTheme extends Theme
+{
+    use Injectable;
+
+    public function asArray()
+    {
+        $colourMap = parent::asArray();
+        $colourMap['cyan'] = 'royalblue';
+        $colourMap['yellow'] = 'goldenrod';
```
Comment (GuySartorelli) [inline src/PolyExecution/AnsiToHtmlTheme.php:19]: The old colours were really hard to see against a white bg

File: src/Security/Confirmation/Handler.php [hunk 14]
```diff
@@ -7,11 +7,6 @@
 use SilverStripe\Control\Director;
 use SilverStripe\Control\HTTPRequest;
 use SilverStripe\Control\RequestHandler;
-use SilverStripe\Forms\Form as BaseForm;
-use SilverStripe\Forms\FieldList;
-use SilverStripe\Forms\TextField;
-use SilverStripe\Forms\FormAction;
-use SilverStripe\Forms\RequiredFields;
```
Comment (GuySartorelli) [inline src/Security/Confirmation/Handler.php:14]: Unrelated refactor - none of these are used.

File: src/View/SSViewer_DataPresenter.php [hunk 15]
```diff
@@ -432,6 +433,11 @@ protected function castValue($value, $source)
             return $value;
         }
 
+        // Wrap list arrays in ViewableData so templates can handle them
+        if (is_array($value) && array_is_list($value)) {
+            return ArrayList::create($value);
+        }
```
Comment (GuySartorelli) [inline src/View/SSViewer_DataPresenter.php]: This was necessary for the params to be passed nicely to template. I've added a test too - though note this will likely be refactored as part of the template layer card anyway.

File: templates/SilverStripe/Dev/DevelopmentAdmin.ss
Comment (GuySartorelli) [inline templates/SilverStripe/Dev/DevelopmentAdmin.ss:1]: HTML used to be handled directly in the controller - this makes it easier to update in modules/projects (like the tasks list template already is) and makes a cleaner separation between logic and presentation.

File: _config/cli.yml [hunk 16]
```diff
@@ -0,0 +1,13 @@
+---
+Name: cli-config
+---
+SilverStripe\Core\Injector\Injector:
+  Symfony\Contracts\EventDispatcher\EventDispatcherInterface.sake:
+    class: 'Symfony\Component\EventDispatcher\EventDispatcher'
+  Symfony\Component\Console\Formatter\OutputFormatterInterface:
+    class: 'Symfony\Component\Console\Formatter\OutputFormatter'
+    calls:
+      - ['setDecorated', [true]]
```
Comment (GuySartorelli) [inline _config/cli.yml:10]: Allows for easy setup of `HtmlOutputFormatter`

File: _config/cli.yml [hunk 17]
```diff
@@ -0,0 +1,13 @@
+---
+Name: cli-config
+---
+SilverStripe\Core\Injector\Injector:
+  Symfony\Contracts\EventDispatcher\EventDispatcherInterface.sake:
+    class: 'Symfony\Component\EventDispatcher\EventDispatcher'
```
Comment (GuySartorelli) [inline _config/cli.yml:6]: Required for Sake event dispatcher, which allows devs to hook into symfony/console events if they want to

File: src/PolyExecution/PolyOutputLogHandler.php
Comment (GuySartorelli) [inline src/PolyExecution/PolyOutputLogHandler.php:1]: This is used by QueuedJobs but seemed sensible to put here so it can be used for any logging anyone wants to do anywhere.

File: src/Cli/Command/NavigateCommand.php [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+            ($responseCode >= 400 && $responseCode < 500) => Command::INVALID,
+            default => Command::FAILURE,
+        };
+    }
+
+    protected function configure(): void
+    {
+        $this->setHelp(<<<HELP
+        Use verbose mode to see the HTTP response status code.
+        The <info>get-var</> arg can either be separated GET variables, or a full query string
```
Comment (emteknetnz) [inline src/Cli/Command/NavigateCommand.php]:
```suggestion
        The <info>get-vars</> arg can either be separated GET variables, or a full query string
```

File: src/Cli/Command/NavigateCommand.php [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+          e.g: <comment>sake navigate about-us/team q=test arrayval[]=value1 arrayval[]=value2</>
+          e.g: <comment>sake navigate about-us/team q=test<info>&</info>arrayval[]=value1<info>&</info>arrayval[]=value2</>
+        HELP);
+        $this->addArgument(
+            'path',
+            InputArgument::REQUIRED,
+            'Relative path to navigate to (e.g: <info>about-us/team</>). Can optionally start with a "/"'
+        );
+        $this->addArgument(
+            'get-var',
```
Comment (emteknetnz) [inline src/Cli/Command/NavigateCommand.php]:
```suggestion
            'get-vars',
```

File: src/Control/CLIRequestBuilder.php [hunk 20]
```diff
@@ -93,4 +93,17 @@ public static function createFromVariables(array $variables, $input, $url = null
...
         return $request;
     }
+
+    public static function createFromInput(InputInterface $input): HTTPRequest
+    {
+        $variables = [];
+        $variables['_SERVER']['argv'] = [
+            'sake',
+            $input->getArgument('path'),
+            ...$input->getArgument('get-var'),
```
Comment (emteknetnz) [inline src/Control/CLIRequestBuilder.php]:
```suggestion
            ...$input->getArgument('get-vars'),
```

File: tests/php/Cli/Command/NavigateCommandTest.php [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+
+    #[DataProvider('provideExecute')]
+    public function testExecute(string $path, array $getVars, int $expectedExitCode, string $expectedOutput): void
+    {
+        // Intentionally override existing rules
+        Director::config()->set('rules', ['test-controller' => TestController::class]);
+        $navigateCommand = new NavigateCommand();
+        $inputParams = [
+            'path' => $path,
+            'get-var' => $getVars,
```
Comment (emteknetnz) [inline tests/php/Cli/Command/NavigateCommandTest.php]:
```suggestion
            'get-vars' => $getVars,
```

File: src/Cli/LegacyParamArgvInput.php (hunk 4 above)
Comment (emteknetnz) [reply src/Cli/LegacyParamArgvInput.php:25]: Keep as is, though there's no reason to not simply to just bump the symfony deps to ^7.1, especially since CMS 6 hasn't even been released. It seems better explicitly not support 7.0 because it makes are maintenance surface slightly lower.

File: src/Cli/LegacyParamArgvInput.php [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+     */
+    private array $argv;
+
+    public function __construct(?array $argv = null, ?InputDefinition $definition = null)
+    {
+        Deprecation::withNoReplacement(
```
Comment (emteknetnz) [inline src/Cli/LegacyParamArgvInput.php]:
```suggestion
        Deprecation::withSuppressedNotice(
```

File: src/Cli/LegacyParamArgvInput.php [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+                    break;
+                }
+            }
+        }
+        if (!empty($convertedFlags)) {
+            // Make sure it's before the end of options (--) signal if there is one.
+            $tokens = ArrayLib::insertBefore($tokens, $convertedFlags, '--', true, true);
+        }
+        if ($hadLegacyParams) {
+            // We only want the warning once regardless of how many params there are.
```
Comment (emteknetnz) [inline src/Cli/LegacyParamArgvInput.php]:
```suggestion
            // We only want the notice once regardless of how many params there are.
```

File: src/PolyExecution/PolyOutput.php [hunk 24]
```diff
@@ -0,0 +1,250 @@
...
+        $this->setDecorated($decorated);
+        $this->setVerbosity($verbosity);
+    }
+
+    /**
+     * Writes messages to the output - but only if we're using the given output format.
+     * Useful for adding explicit HTML markup (divs, etc) to wrap the main output.
+     * If including HTML markup, use OUTPUT_RAW in the options.
+     *
+     * @param string $listType One of the LIST_* consts, e.g. PolyOutput::LIST_UNORDERED
```
Comment (emteknetnz) [inline src/PolyExecution/PolyOutput.php]: This param doesn't exist

File: src/PolyExecution/PolyOutput.php [hunk 25]
```diff
@@ -0,0 +1,250 @@
...
+     * @param int $options A bitmask of options (one of the OUTPUT or VERBOSITY constants),
+     * 0 is considered the same as self::OUTPUT_NORMAL | self::VERBOSITY_NORMAL
+     */
+    public function writeForFormat(
```
Comment (emteknetnz) [inline src/PolyExecution/PolyOutput.php]: Have this as a single method makes for some fairly un-concise calls e.g.

```php
        $output->writeForFormat(
            PolyOutput::FORMAT_HTML,
            '<pre>',
            options: PolyOutput::OUTPUT_RAW
        );
```

This should be split into separate methods

```php
public function writeForHtml(string|iterable $messages, bool $newline = false, int $options = OutputInterface::OUTPUT_RAW): void

public function writeForAnsi(string|iterable $messages, bool $newline = false, int $options = OutputInterface::OUTPUT_NORMAL): void
```

Which means the above example would simply be

```php
$output->writeForHtml('<pre>');
```

File: src/Dev/Command/DbCleanup.php [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+            foreach ($subclasses as $k => $subclass) {
+                if (!DataObject::getSchema()->classHasTable($subclass)) {
+                    unset($subclasses[$k]);
+                }
+            }
+
+            if ($subclasses) {
+                $records = DB::query("SELECT * FROM \"$baseTable\"");
+
+
```
Comment (emteknetnz) [inline src/Dev/Command/DbCleanup.php]:
```suggestion
```

File: src/PolyExecution/AnsiToHtmlConverter.php (hunk 12 above)
Comment (emteknetnz) [reply src/PolyExecution/AnsiToHtmlConverter.php:69]: Makes sense to keep in I think to understand why this exists

File: src/PolyExecution/HttpRequestInput.php [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+    /**
+     * @param array<InputOption> $commandOptions Any options that apply for the command itself.
+     * Do not include global options (e.g. flush) - they are added explicitly in the constructor.
+     */
+    public function __construct(HTTPRequest $request, array $commandOptions = [])
+    {
+        $definition = new InputDefinition([
+            // Also add global options that are applicable for HTTP requests
+            new InputOption('quiet', null, InputOption::VALUE_NONE, 'Do not output any message'),
+            new InputOption('verbose', null, InputOption::VALUE_OPTIONAL, 'Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug'),
```
Comment (emteknetnz) [inline src/PolyExecution/HttpRequestInput.php:32]: Does this match how symfony console apps normally do verbosity? Composer does -v / -vv / -vvv - is that more standard? --verbose=2 reads a little weird to me

File: src/PolyExecution/PolyCommand.php [hunk 28]
```diff
@@ -0,0 +1,182 @@
...
+        if (Director::isDev()) {
+            return true;
+        }
+        if (!static::config()->get('can_run_in_browser')) {
+            return false;
+        }
+        // Check permissions if there are any
+        $permissions = static::config()->get('permissions_for_browser_execution');
+        if ($permissions) {
+            return Permission::check($permissions);
```
Comment (emteknetnz) [inline src/PolyExecution/PolyCommand.php]: By default this will do an 'any' check, which is different from how LeftAndMain::required_permission_codes does it which does it as an 'all' (not that any of our LeftAndMain subclasses have multiple codes)

Is that what we want?

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: src/Dev/BuildTask.php [hunk 29]
```diff
@@ -2,103 +2,112 @@
...
 
-    /**
-     * @var string $title Shown in the overview on the {@link TaskRunner}
-     * HTML or CLI interface. Should be short and concise, no HTML allowed.
-     */
-    protected $title;
+    private static array $permissions_for_browser_execution = [
+        'ADMIN',
+        'anyone_with_dev_admin_permissions' => 'ALL_DEV_ADMIN',
+        'anyone_with_task_permissions' => 'BUILDTASK_CAN_RUN',
```
Comment (emteknetnz) [inline src/Dev/BuildTask.php]:
```suggestion
        'ALL_DEV_ADMIN',
        'BUILDTASK_CAN_RUN',
```
I found the additional of 'helpful' keys just confusing, I thought they were adding some additional functionality when they're not

File: src/Dev/Command/ConfigAudit.php [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+ * Command to audit the configuration.
+ * Can be run either via an HTTP request or the CLI.
+ */
+class ConfigAudit extends DevCommand
+{
+    protected static string $commandName = 'config:audit';
+
+    protected static string $description = 'Find configuration properties that are not defined (or inherited) by their respective classes';
+
+    private static string|array $permissions_for_browser_execution = [
```
Comment (emteknetnz) [inline src/Dev/Command/ConfigAudit.php]:
```suggestion
    private static array $permissions_for_browser_execution = [
```

File: src/Dev/Command/DevCommand.php [hunk 31]
```diff
@@ -0,0 +1,58 @@
...
+use Symfony\Component\Console\Terminal;
+
+/**
+ * A command that can be run from CLI or via an HTTP request in a dev/* route
+ */
+abstract class DevCommand extends PolyCommand
+{
+    private static array $permissions_for_browser_execution = [
+        'ADMIN',
+        'anyone_with_dev_admin_permissions' => 'ALL_DEV_ADMIN',
```
Comment (emteknetnz) [inline src/Dev/Command/DevCommand.php]:
```suggestion
        'ALL_DEV_ADMIN',
```
I found the additional of 'helpful' keys just confusing, I thought they were adding some additional functionality when they're not

File: src/Dev/Tasks/CleanupTestDatabasesTask.php [hunk 32]
```diff
@@ -2,46 +2,32 @@
...
+    protected string $title = 'Deletes all temporary test databases';
 
-    protected $title = 'Deletes all temporary test databases';
+    protected static string $description = 'Cleans up leftover databases from aborted test executions (starting with ss_tmpdb)';
 
-    protected $description = 'Cleans up leftover databases from aborted test executions (starting with ss_tmpdb)';
+    private static string|array|null $permissions_for_browser_execution = [
+        'anyone_with_dev_admin_permissions' => null,
+        'anyone_with_task_permissions' => null,
+    ];
```
Comment (emteknetnz) [inline src/Dev/Tasks/CleanupTestDatabasesTask.php]:
```suggestion
    public static function canRunInBrowser(): bool
    {
        if (!parent::canRunInBrowser()) {
            return false;
        }
        return Permission::check('ADMIN');
    }
```
This is a much cleaner way to do the 'must be admin' check rather than overriding the default 'helpful' keys

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: src/Cli/Command/NavigateCommand.php (hunk 18 above)
Comment (GuySartorelli) [reply src/Cli/Command/NavigateCommand.php]: Done - also updated in `CLIRequestBuilder::createFromInput()`

File: src/Cli/Command/NavigateCommand.php (hunk 19 above)
Comment (GuySartorelli) [reply src/Cli/Command/NavigateCommand.php]: Done

File: src/Control/CLIRequestBuilder.php (hunk 20 above)
Comment (GuySartorelli) [reply src/Control/CLIRequestBuilder.php]: Done

File: tests/php/Cli/Command/NavigateCommandTest.php (hunk 21 above)
Comment (GuySartorelli) [reply tests/php/Cli/Command/NavigateCommandTest.php]: Done

File: src/Cli/LegacyParamArgvInput.php (hunk 22 above)
Comment (GuySartorelli) [reply src/Cli/LegacyParamArgvInput.php]: That stuff hasn't been merged up yet.

File: src/Cli/LegacyParamArgvInput.php (hunk 23 above)
Comment (GuySartorelli) [reply src/Cli/LegacyParamArgvInput.php]: It's a warning. I don't know why our API says "notice" everywhere but it's logged as a warning.

File: src/PolyExecution/PolyOutput.php (hunk 24 above)
Comment (GuySartorelli) [reply src/PolyExecution/PolyOutput.php]: Oops, copy/paste error. Removed.

File: src/PolyExecution/PolyOutput.php (hunk 25 above)
Comment (GuySartorelli) [reply src/PolyExecution/PolyOutput.php]: Done - I considered not making `writeForHtml()` raw by default but I've decided to just clarify that in the docs and PHPDoc instead. Hopefully people look before they leap.

File: src/Dev/Command/DbCleanup.php (hunk 26 above)
Comment (GuySartorelli) [reply src/Dev/Command/DbCleanup.php]: Done

File: src/PolyExecution/PolyCommand.php (hunk 28 above)
Comment (GuySartorelli) [reply src/PolyExecution/PolyCommand.php]: This is copied over from the now deprecated `init_permissions` config property that existed on the controllers.
The idea is that you can add finer grain control (admins have access to all commands, people who have "all dev" have access to al dev commands, and people with specific permissions for individual commands have access to just those specific commands).

File: src/Dev/BuildTask.php (hunk 29 above)
Comment (GuySartorelli) [reply src/Dev/BuildTask.php]: They _are_ adding additional functionality - you can set these to `null` in subclasses if you don't want people with those permissions to access a specific task. You might have a task that's highly destructive that only admins can run, for example.

File: src/Dev/Command/ConfigAudit.php (hunk 30 above)
Comment (GuySartorelli) [reply src/Dev/Command/ConfigAudit.php]: Good catch, done. There was an older iteration where I did allow string - but I decided later that's silly.

File: src/Dev/Command/DevCommand.php (hunk 31 above)
Comment (GuySartorelli) [reply src/Dev/Command/DevCommand.php]: See https://github.com/silverstripe/silverstripe-framework/pull/11353#discussion_r1772353673

File: src/Dev/Tasks/CleanupTestDatabasesTask.php (hunk 32 above)
Comment (GuySartorelli) [reply src/Dev/Tasks/CleanupTestDatabasesTask.php]: The suggested change makes it not configurable.
The intention is that it's configurable.

Comment (emteknetnz) [reply src/Dev/Tasks/CleanupTestDatabasesTask.php]: Fair enough, though I think this is a really clunky and un-intuitive way handle permissions. I didn't really grasp what this was for and I was spending a bunch of time peer reviewing this code. The average developer who only glances at this is most likely to not understand what's going on

Is there anywhere else in the core code base we have this pattern of using keys for permissions so they can be toggled off?

Instead, perhaps just add a `private static $admin_only_for_browser_execution = false;` on the base class and just set it to true on this job and have that override 'permissions_for_browser_execution'

Comment (GuySartorelli) [reply src/Dev/Tasks/CleanupTestDatabasesTask.php]: > Is there anywhere else in the core code base we have this pattern of using keys for permissions so they can be toggled off?

Dunno, but we do use this pattern for other things that aren't permissions.

At the end of the day this is the pre-existing pattern that the permissions for these exact parts of the code were already using. I've just moved it, it already existed. I don't want to change it as part of this card.

Would it help if I added a comment to the PHPDoc about the fact that setting a permission to `null` will skip checking that permission?

Comment (emteknetnz) [reply src/Dev/Tasks/CleanupTestDatabasesTask.php]: Yeah, the `anyone_with_dev_admin_permissions` and `anyone_with_task_permissions` keys aren't existing though

Does this work? If not would it be feasible to make this work?

'ADMIN',
'ALL_DEV_ADMIN' => true,
'BUILDTASK_CAN_RUN' => true,

Then set to false in this particular job?

That's probably ideal. Another options would be to make the keys match the permissions

'ADMIN',
'all_dev_admin' => 'ALL_DEV_ADMIN',
'buildtask_can_run' => 'BUILDTASK_CAN_RUN',

or

'ADMIN',
'ALL_DEV_ADMIN' => 'ALL_DEV_ADMIN',
'BUILDTASK_CAN_RUN' => 'BUILDTASK_CAN_RUN',

Comment (GuySartorelli) [reply src/Dev/Tasks/CleanupTestDatabasesTask.php]: Yeah that's more intuitive, I can make that work. As discussed in person the top one is preferred so I'll run with that.

Comment (GuySartorelli) [reply src/Dev/Tasks/CleanupTestDatabasesTask.php]: Done

File: src/Cli/LegacyParamArgvInput.php (hunk 22 above)
Comment (emteknetnz) [reply src/Cli/LegacyParamArgvInput.php]: Should rebase it in then so we don't have things break later

File: src/Cli/LegacyParamArgvInput.php (hunk 23 above)
Comment (emteknetnz) [reply src/Cli/LegacyParamArgvInput.php]: It's an `E_USER_DEPRECATED`? Which is neither. Since we say 'notice' everywhere we should just refer to it as a notice rather than a warning.

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: src/Cli/LegacyParamArgvInput.php (hunk 23 above)
Comment (GuySartorelli) [reply src/Cli/LegacyParamArgvInput.php]: > It's an `E_USER_DEPRECATED`?

Yes, and it's logged as a warning.
We do actually use both "warning" and "notice" to refer to it depending on where you look - this isn't the only place that does it.

It doesn't really matter, it's a comment and we all know what it means. I'm gonna change it at this stage just because arguing about this is a waste of everyone's time.

File: src/Cli/LegacyParamArgvInput.php (hunk 22 above)
Comment (GuySartorelli) [reply src/Cli/LegacyParamArgvInput.php]: It hadn't been merged up from 5 to 6. I can't rebase on top of something that isn't in the branch I need to rebase on??
Since those comments were made it _has_ been merged up, so _now_ I'm able to make the change. Though it really shouldn't have been blocked for merging for something that wasn't even in the branch I'm targetting 😅

Comment (GuySartorelli) [reply src/Cli/LegacyParamArgvInput.php]: Done

**Review by emteknetnz — APPROVED** [review]
